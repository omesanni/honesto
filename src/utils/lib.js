import React from 'react';
import AsyncComponent from '../components/AsyncComponent';

/**
 * Lazy loads a component
 * @param  {Function} getComponent Returns a dynamic import of the component
 * @param  {Object}   props
 * @return {ReactElement}
 */
export function lazyLoadComponent(getComponent, props = {}) {
  return React.createElement(AsyncComponent, { ...props, render: getComponent });
}
