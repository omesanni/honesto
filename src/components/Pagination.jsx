import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { setRange, isWithinRange } from '../utils/lib';

const Pagination = (props) => {
  const { current, pages } = props;
  const range = setRange(current, props.spread, pages);

  // rendering logic
  const show = {
    first: () => (current !== 1 && range.min !== 1),
    prev: () => (current !== 1),
    next: () => (current < pages),
    last: () => (current !== pages && range.max !== pages),
  };

  // render a range of pagination numbers
  function renderButtons() {
    const buttons = [];

    for (let i = range.min; isWithinRange(range, i, pages); i += 1) {
      buttons.push(
        <div
          className={classnames('page-item', {
            active: i === current,
          })}
          key={i}
          onClick={() => props.onChange(i)}
        >
          <span className={'page-link'}>
            {i}
          </span>
        </div>
      );
    }

    return buttons.length > 1 ? buttons : null;
  }

  return (
    <nav
      className={classnames('d-flex justify-content-center', props.className)}
    >
      <div className={'pagination'}>
        <div
          className={classnames('page-item', {
            'disabled no-pointer-events': !show.prev(),
          })}
          onClick={() => props.onChange(current - 1)}
        >
          <span className={'page-link'}>
            {'Prev'}
          </span>
        </div>

        {show.first() ? (
          <div
            className={'page-item'}
            onClick={() => props.onChange(1)}
          >
            <span className={'page-link'}>
              {1}
            </span>
          </div>
        ) : null}

        {show.first() ? (
          <div
            className={'page-item disabled'}
          >
            <span className={'page-link'}>
              {'...'}
            </span>
          </div>
        ) : null}

        {renderButtons()}

        {show.last() ? (
          <div
            className={'page-item disabled'}
          >
            <span className={'page-link'}>
              {'...'}
            </span>
          </div>
        ) : null}

        {show.last() ? (
          <div
            className={'page-item'}
            onClick={() => props.onChange(pages)}
          >
            <span className={'page-link'}>
              {pages}
            </span>
          </div>
        ) : null}

        <div
          className={classnames('page-item', {
            'disabled no-pointer-events': !show.next(),
          })}
          onClick={() => props.onChange(current + 1)}
        >
          <span className={'page-link'}>
            {'Next'}
          </span>
        </div>
      </div>
    </nav>
  );
};

Pagination.propTypes = {
  current: PropTypes.number.isRequired,
  onChange: PropTypes.func,
  pages: PropTypes.number.isRequired,
  spread: PropTypes.number,
  className: PropTypes.string,
};

Pagination.defaultProps = {
  spread: 4,
  onChange: () => undefined,
  className: '',
};

export default Pagination;
