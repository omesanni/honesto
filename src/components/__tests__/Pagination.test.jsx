import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../Pagination';
import { expectToMatchSnapshot } from '../../utils/tests';

describe('Pagination', () => {
  let component;

  const props = {
    current: 10,
    pages: 25,
    onChange: jest.fn(),
  };

  function getPagination(p = props) {
    return <Pagination {...p} />;
  }

  function mountComponent(ownProps = props) {
    return shallow(getPagination(ownProps));
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders with ellipsis after page 1 and before last page', () => {
    expectToMatchSnapshot(getPagination());
  });

  it('renders with ellipsis only before last page', () => {
    expectToMatchSnapshot(getPagination({ ...props, current: 1 }));
  });

  it('renders with ellipsis only after page 1', () => {
    expectToMatchSnapshot(getPagination({ ...props, current: 24 }));
  });

  describe('onChange()', () => {
    it('returns preceding page when previous button is clicked', () => {
      component = mountComponent();

      component.find('.page-item').first().simulate('click');

      expect(props.onChange).toHaveBeenCalledWith(props.current - 1);
    });

    it('returns next page when next button is clicked', () => {
      component = mountComponent();

      component.find('.page-item').last().simulate('click');

      expect(props.onChange).toHaveBeenCalledWith(props.current + 1);
    });

    it('should change page when a page within spread range is clicked', () => {
      component = mountComponent();
      const page = component.find('.page-item').findWhere(item => item.key()).at(0);

      page.simulate('click');

      expect(props.onChange).toHaveBeenCalledWith(Number(page.key()));
    });

    it('should go to first page', () => {
      component = mountComponent();

      component.find('.page-item').at(1).simulate('click');

      expect(props.onChange).toHaveBeenCalledWith(1);
    });

    it('should go to last page', () => {
      component = mountComponent();
      const pageItems = component.find('.page-item');

      pageItems.at(pageItems.length - 2).simulate('click');

      expect(props.onChange).toHaveBeenCalledWith(props.pages);
    });
  });
});
