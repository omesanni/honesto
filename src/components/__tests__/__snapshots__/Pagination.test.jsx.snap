// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders with ellipsis after page 1 and before last page 1`] = `
<nav
  className="d-flex justify-content-center"
>
  <div
    className="pagination"
  >
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Prev
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        1
      </span>
    </div>
    <div
      className="page-item disabled"
    >
      <span
        className="page-link"
      >
        ...
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        6
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        7
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        8
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        9
      </span>
    </div>
    <div
      className="page-item active"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        10
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        11
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        12
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        13
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        14
      </span>
    </div>
    <div
      className="page-item disabled"
    >
      <span
        className="page-link"
      >
        ...
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        25
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Next
      </span>
    </div>
  </div>
</nav>
`;

exports[`Pagination renders with ellipsis only after page 1 1`] = `
<nav
  className="d-flex justify-content-center"
>
  <div
    className="pagination"
  >
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Prev
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        1
      </span>
    </div>
    <div
      className="page-item disabled"
    >
      <span
        className="page-link"
      >
        ...
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        17
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        18
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        19
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        20
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        21
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        22
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        23
      </span>
    </div>
    <div
      className="page-item active"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        24
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        25
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Next
      </span>
    </div>
  </div>
</nav>
`;

exports[`Pagination renders with ellipsis only before last page 1`] = `
<nav
  className="d-flex justify-content-center"
>
  <div
    className="pagination"
  >
    <div
      className="page-item disabled no-pointer-events"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Prev
      </span>
    </div>
    <div
      className="page-item active"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        1
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        2
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        3
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        4
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        5
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        6
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        7
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        8
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        9
      </span>
    </div>
    <div
      className="page-item disabled"
    >
      <span
        className="page-link"
      >
        ...
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        25
      </span>
    </div>
    <div
      className="page-item"
      onClick={[Function]}
    >
      <span
        className="page-link"
      >
        Next
      </span>
    </div>
  </div>
</nav>
`;
